File: 03 Statistical Diagnosis of Financial Data/Normality Test.R
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

################################################################################
d <- read.csv("../Datasets/stock_1999_2002.csv", row.names=1)	  # read in data file
d <- as.ts(d)
u <- (lag(d) - d) / d
colnames(u) <- colnames(d)

library(zoo)
plot(zoo(d), plot.type="multiple", col=c("blue", "orange", "green"))
plot(zoo(u), plot.type="multiple", col=c("blue", "orange", "green"))

par(mfrow=c(3,2), mar=c(4,4,4,4))
# if the dist is normal, the plot should close to this line
# histogram; qq-normal plot; add a line for reference
hist(u[,"HSBC"]); qqnorm(u[,"HSBC"]); qqline(u[,"HSBC"])
hist(u[,"CLP"]); qqnorm(u[,"CLP"]); qqline(u[,"CLP"])
hist(u[,"CK"]); qqnorm(u[,"CK"]); qqline(u[,"CK"])

################################################################################
shapiro.test(u[,"HSBC"])
shapiro.test(u[,"CLP"])
shapiro.test(u[,"CK"])

################################################################################
u1 <- u[,"HSBC"]; u2 <- u[,"CLP"]; u3 <- u[,"CK"]
ks.test(u1, pnorm, mean=mean(u1), sd=sd(u1))
ks.test(u2, pnorm, mean=mean(u2), sd=sd(u2))
ks.test(u3, pnorm, mean=mean(u3), sd=sd(u3))

################################################################################
library("tseries")

JB.test <- function(u){
  z <- u - mean(u)            # Remove mean
  n <- length(z)              # Sample size
  s <- sd(z)*sqrt((n-1)/n)	# Population standard deviation
  sk <- sum(z^3)/(n*s^3)      # Skewness
  ku <- sum(z^4)/(n*s^4) - 3  # Excess Kurtosis
  JB <- n * (sk^2/6 + ku^2/24)	# JB test statistics
  p <- 1 - pchisq(JB, 2)      # chi-squared p-value
  list("JB_stat"=JB, "p_value"=p)
}

JB.test(u[,"HSBC"])
jarque.bera.test(u[,"HSBC"])

JB.test(u[,"CLP"])
jarque.bera.test(u[,"CLP"])

JB.test(u[,"CK"])
jarque.bera.test(u[,"CK"])

################################################################################
library("car")

t.QQ.plot <- function(u, comp=""){
  z <- u - mean(u) # Remove mean
  sz <- sort(z)		   # sort z
  n <- length(z)		  # sample size
  s <- sd(z)*sqrt((n-1)/n)	# Population standard deviation
  ku <- sum(z^4)/(n*s^4) - 3	# Excess kurtosis
  nu <- 6/ku + 4 # Degrees of freedom
  i <- ((1:n)-0.5)/n  # create a vector of percentiles
  q <- qt(i, nu)		  # percentile points from t(v)

  plot(q, sz, main=paste("Self-defined t Q-Q Plot of ", comp, " Return"))
  qqline(sz, distribution=function(p) qt(p, df=nu), probs=c(0.25, 0.75))
  qqPlot(z, distribution="t", df=nu, envelope=FALSE, line="quartiles",
         col.lines="black", lwd=1, cex=1, grid=FALSE, id=FALSE,
         main=paste("t Q-Q Plot of ", comp, " Return"))
  nu
}

par(mfrow=c(3,2), mar=c(4,4,4,4))
df_HSBC <- t.QQ.plot(u[,"HSBC"], comp="HSBC")
df_CLP <- t.QQ.plot(u[,"CLP"], comp="CLP")
df_CK <- t.QQ.plot(u[,"CK"], comp="CK")

################################################################################
t_HSBC <- u[,"HSBC"]/sd(u[,"HSBC"])*sqrt(df_HSBC/(df_HSBC-2))
ks.test(t_HSBC, pt, df_HSBC)
t_CLP <- u[,"CLP"]/sd(u[,"CLP"])*sqrt(df_CLP/(df_CLP-2))
ks.test(t_CLP, pt, df_CLP)
t_CK <- u[,"CK"]/sd(u[,"CK"])*sqrt(df_CK/(df_CK-2))
ks.test(t_CK, pt, df_CK)

################################################################################
n <- 180
u_180 <- tail(u, n)
mu_180 <- apply(u_180, 2, mean)
S_180 <- cov(u_180)

z_180 <- sweep(u_180, 2, mu_180)
md2_180 <- rowSums((z_180 %*% solve(S_180)) * z_180)
smd2_180 <- sort(md2_180)		# sort md2 in ascendingly
i <- ((1:n)-0.5)/n		# create percentile vector
q <- qchisq(i,3)		# compute quantiles 

par(mfrow=c(1,1))
qqplot(q, smd2_180, main="Chi2 Q-Q Plot")		# QQ-chisquare plot
qqline(smd2_180, distribution=function(p) qchisq(p, df=3))

ks.test(smd2_180, pchisq, 3)

################################################################################
cor(u_180)

################################################################################
pairs(u_180)

################################################################################
par(mfrow=c(4,3), mar=c(4,4,4,4))

hist(d[,"HSBC"]); hist(d[,"CLP"]); hist(d[,"CK"])

qqnorm(d[,"HSBC"]); qqline(d[,"HSBC"])
qqnorm(d[,"CLP"]); qqline(d[,"CLP"])
qqnorm(d[,"CK"]); qqline(d[,"CK"])

plot(d[,"HSBC"], lag(d[,"HSBC"]))
plot(d[,"CLP"], lag(d[,"CLP"]))
plot(d[,"CK"], lag(d[,"CK"]))

plot(u[,"HSBC"], lag(u[,"HSBC"]))
plot(u[,"CLP"], lag(u[,"CLP"]))
plot(u[,"CK"], lag(u[,"CK"]))

################################################################################
par(mfrow=c(3,3), mar=c(4,4,4,4))

acf(d[,"HSBC"]); acf(d[,"CLP"]); acf(d[,"CK"])
acf(u[,"HSBC"]); acf(u[,"CLP"]); acf(u[,"CK"])
acf(u[,"HSBC"]^2); acf(u[,"CLP"]^2); acf(u[,"CK"]^2)

################################################################################
set.seed(4002)

mu_180 <- apply(u_180, 2, mean)
S_180 <- cov(u_180)
C_180 <- chol(S_180)  # Cholesky decomposition of Sigma
s0 <- tail(d, 1)      # set s0 to the most recent price
s_pred <- c()
for (i in 1:90) {
  z <- rnorm(3)
  v <- mu_180 + t(C_180) %*% z
  s1 <- s0 * (1 + t(v))	# new stock price
  s_pred <- rbind(s_pred, s1)
  s0 <- s1	# update s0
}

s_pred <- ts(s_pred, start=nrow(d)+1)
data <- ts.union(d, s_pred)
par(mfrow=c(1,1))

col <- c("blue", "orange", "green", "pink", "brown", "red")
plot(data, plot.type="s", col=col)
legend("topright", col=col, lty=1,
       legend=c(colnames(d), paste0(colnames(d), "_pred")))

File: 03 Statistical Diagnosis of Financial Data/Normality Test.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

d = pd.read_csv("stock_1999_2002.csv")
u = np.diff(d, axis=0) / d.iloc[:-1, :] # Arithmetic return

d.plot(subplots=True, layout=(3,1), figsize=(10, 10))
u.plot(subplots=True, layout=(3,1), sharey=True, figsize=(10, 10))

fig, axs = plt.subplots(3, 2, figsize=(10,15))
axs[0,0].hist(u.HSBC, color="blue", ec='black', bins="sturges")
axs[0,0].set_title("Histogram of HSBC Return")
stats.probplot(u.HSBC, dist="norm", plot=axs[0,1])
axs[0,1].set_title("Normal Q-Q Plot of HSBC Return")
axs[0,1].get_lines()[0].set_color('blue')

axs[1,0].hist(u.CLP, color="orange", ec='black', bins="sturges")
axs[1,0].set_title("Histogram of CLP Return")
stats.probplot(u.CLP, dist="norm", plot=axs[1,1])
axs[1,1].set_title("Normal Q-Q Plot of CLP Return")
axs[1,1].get_lines()[0].set_color('orange')

axs[2,0].hist(u.CK, color="green", ec='black', bins="sturges")
axs[2,0].set_title("Histogram of CK Return")
stats.probplot(u.CK, dist="norm", plot=axs[2,1])
axs[2,1].set_title("Normal Q-Q Plot of CK Return")
axs[2,1].get_lines()[0].set_color('green')
fig.tight_layout()

#%%

print(stats.shapiro(u.HSBC))
print(stats.shapiro(u.CLP))
print(stats.shapiro(u.CK))

#%%
print(stats.kstest(u.HSBC, stats.norm.cdf, 
                   args=(np.mean(u.HSBC), np.std(u.HSBC))))
print(stats.kstest(u.CLP, stats.norm.cdf,
                   args=(np.mean(u.CLP), np.std(u.CLP))))
print(stats.kstest(u.CK, stats.norm.cdf,
                   args=(np.mean(u.CK), np.std(u.CK))))

#%%

def JB_test(u):
    z = u - np.mean(u)  # Remove mean
    n = len(z) # Sample size
    s = np.std(z) # Population standard deviation
    sk = sum(z**3) / (n*s**3) # Skewness
    ku = sum(z**4) / (n*s**4) - 3 # Excess Kurtosis
    JB = n * (sk**2/6 + ku**2/24) # JB test statistics
    p = 1 - stats.chi2.cdf(JB, 2) # chi-squared p-value
    return ({"JB-test": JB, "p-value": p})

print(stats.jarque_bera(u.HSBC))
print(JB_test(u.HSBC))

print(stats.jarque_bera(u.CLP))
print(JB_test(u.CLP))

print(stats.jarque_bera(u.CK))
print(JB_test(u.CK))

#%%

def QQt_plot(u, color="blue", comp="", ax=None):
    z = u - np.mean(u)  # Remove mean
    sz = np.sort(z) # Sort z
    n = len(z) # Sample size
    s = np.std(z) # Population standard deviation
    ku = sum(z**4) / (n*s**4) - 3 # Excess Kurtosis
    nu = 6/ku + 4 # degrees of freedom
    i = (np.arange(1, n+1)-0.5)/n # create a vector of percentile
    q = stats.t.ppf(i, nu)
    
    b, w = np.linalg.lstsq(np.vstack([np.ones(n), q]).T, sz, 
                           rcond=None)[0]
    ax.scatter(q, sz, color=color)
    ax.plot(q, w*q + b, color="red")
    ax.set_title(f"Self-defined t Q-Q Plot of {comp} Return")
    return nu

fig, axs = plt.subplots(3, 2, figsize=(10,15))
df_HSBC = QQt_plot(u.HSBC, color="blue", comp="HSBC", ax=axs[0,0])
stats.probplot(u.HSBC, dist="t", sparams=df_HSBC, plot=axs[0,1])
axs[0,1].set_title("t Q-Q Plot of HSBC Return")
axs[0,1].get_lines()[0].set_color('blue')

df_CLP = QQt_plot(u.CLP, color="orange", comp="CLP", ax=axs[1,0])
stats.probplot(u.CLP, dist="t", sparams=df_CLP, plot=axs[1,1])
axs[1,1].set_title("t Q-Q Plot of CLP Return")
axs[1,1].get_lines()[0].set_color('orange')

df_CK = QQt_plot(u.CK, color="green", comp="CK", ax=axs[2,0])
stats.probplot(u.CK, dist="t", sparams=df_CK, plot=axs[2,1])
axs[2,1].set_title("t Q-Q Plot of CK Return")
axs[2,1].get_lines()[0].set_color('green')
fig.tight_layout()

#%%
print([df_HSBC, df_CLP, df_CK])

print(stats.kstest(u.HSBC, stats.t.cdf, args=(df_HSBC,)))
print(stats.kstest(u.CLP, stats.t.cdf, args=(df_CLP,)))
print(stats.kstest(u.CK, stats.t.cdf, args=(df_CK,)))

#%%
n = 180
u_180 = u.iloc[len(u)-n:, :]
mu_180 = np.mean(u_180)
S_180 = np.cov(u_180, rowvar=False)

z_180 = (u_180 - mu_180).values.reshape(n, -1)
d2_180 = np.diagonal(z_180 @ np.linalg.inv(S_180) @ z_180.T)
sd2_180 = np.sort(d2_180)
i = (np.arange(1, n+1)-0.5)/n
q = stats.chi2.ppf(i, 3)

print(mu_180)
print(z_180)
print(d2_180)

fig = plt.figure()
b, w = np.linalg.lstsq(np.vstack([np.ones(n), q]).T, sd2_180, 
                       rcond=None)[0]
plt.scatter(q, sd2_180, color="blue")
plt.plot(q, w*q + b, color="blue")
plt.title("Chi2 Q-Q Plot")

print(stats.kstest(d2_180, stats.chi2.cdf, args=(3,)))

#%%

print(np.corrcoef(u_180, rowvar=False))

#%%
hist_kwds = {'color':'blue', 'bins':'sturges', 'ec':'black'}
axes = pd.plotting.scatter_matrix(u, figsize=(10,10), color="blue",
                                  hist_kwds=hist_kwds)
new_labels = [round(float(i.get_text()), 2) for i in 
              axes[0,0].get_yticklabels()]
axes[0,0].set_yticklabels(new_labels)
plt.tight_layout()

#%%

fig, axs = plt.subplots(4, 3, figsize=(15,20))
axs[0,0].hist(d.HSBC, color="blue", ec='black', bins="sturges")
axs[0,0].set_title("Histogram of HSBC Price")
stats.probplot(d.HSBC, dist="norm", plot=axs[1,0])
axs[1,0].set_title("Normal Q-Q Plot of HSBC Price")
axs[1,0].get_lines()[0].set_color('blue')
pd.plotting.lag_plot(d.HSBC, lag=1, ax=axs[2,0], c="blue")
axs[2,0].set_title("1-day Lagged Plot of HSBC Price")
pd.plotting.lag_plot(u.HSBC, lag=1, ax=axs[3,0], c="blue")
axs[3,0].set_title("1-day Lagged Plot of HSBC Return")

axs[0,1].hist(d.CLP, color="orange", ec='black', bins="sturges")
axs[0,1].set_title("Histogram of CLP Price")
stats.probplot(d.CLP, dist="norm", plot=axs[1,1])
axs[1,1].set_title("Normal Q-Q Plot of CLP Price")
axs[1,1].get_lines()[0].set_color('orange')
pd.plotting.lag_plot(d.CLP, lag=1, ax=axs[2,1], c="orange")
axs[2,1].set_title("1-day Lagged Plot of CLP Price")
pd.plotting.lag_plot(u.CLP, lag=1, ax=axs[3,1], c="orange")
axs[3,1].set_title("1-day Lagged Plot of CLP Return")

axs[0,2].hist(d.CK, color="green", ec='black', bins="sturges")
axs[0,2].set_title("Histogram of CK Price")
stats.probplot(d.CK, dist="norm", plot=axs[1,2])
axs[1,2].set_title("Normal Q-Q Plot of CK Price")
axs[1,2].get_lines()[0].set_color('green')
pd.plotting.lag_plot(d.CK, lag=1, ax=axs[2,2], c="green")
axs[2,2].set_title("1-day Lagged Plot of CK Price")
pd.plotting.lag_plot(u.CK, lag=1, ax=axs[3,2], c="green")
axs[3,2].set_title("1-day Lagged Plot of CK Return")
plt.tight_layout()

#%%

from statsmodels.graphics.tsaplots import plot_acf
fig, axs = plt.subplots(3, 3, figsize=(15,15))
plot_acf(d.HSBC, alpha=None, c="blue", title="ACF Plot of HSBC Price", ax=axs[0,0])
plot_acf(u.HSBC, alpha=None, c="blue", title="ACF Plot of HSBC Return", ax=axs[1,0])
plot_acf(u.HSBC**2, alpha=None, c="blue", title="ACF Plot of Squared HSBC Return", ax=axs[2,0])

plot_acf(d.CLP, alpha=None, c="orange", title="ACF Plot of CLP Price", ax=axs[0,1])
plot_acf(u.CLP, alpha=None, c="orange", title="ACF Plot of CLP Return", ax=axs[1,1])
plot_acf(u.CLP**2, alpha=None, c="orange", title="ACF Plot of Squared CLP Return", ax=axs[2,1])

plot_acf(d.CK, alpha=None, c="green", title="ACF Plot of CK Price", ax=axs[0,2])
plot_acf(u.CK, alpha=None, c="green", title="ACF Plot of CK Return", ax=axs[1,2])
plot_acf(u.CK**2, alpha=None, c="green", title="ACF Plot of Squared CK Return", ax=axs[2,2])

#%%
np.random.seed(4012)
mu_180 = np.mean(u_180)
S_180 = np.cov(u_180, rowvar=False)
C_180 = np.linalg.cholesky(S_180)
s0 = d.iloc[-1,:]
s_pred = []
for i in range(90):
    z = np.random.randn(3)
    v = mu_180 + C_180.T @ z
    s1 = s0 * (1 + v)
    s_pred.append(s1.values)
    s0 = s1

df_pred = pd.DataFrame(np.array(s_pred), columns=d.columns.values + "_pred")
df_pred.index = np.arange(len(d)+1, len(d)+90+1)

pd.merge(d, df_pred, how='outer', left_index=True, 
         right_index=True).plot(figsize=(10, 7))

